graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/SmartRotationStrategy.java

```diff
@@ -50,14 +66,83 @@ public RotationStrategyConfig defaultConfiguration() {
         return SmartRotationStrategyConfig.builder().build();
     }
 
-    @Nullable
     @Override
+    @Nonnull
     protected Result shouldRotate(final String index, IndexSet indexSet) {
-        return null;
+        final DateTime creationDate = indices.indexCreationDate(index).orElseThrow(()-> new IllegalStateException("No index creation date"));
+        final Long sizeInBytes = indices.getStoreSizeInBytes(index).orElseThrow(() -> new IllegalStateException("No index size"));
+
+        Period leeWay;
+        if (indexSet.getConfig().rotationStrategy() instanceof SmartRotationStrategyConfig rotationConfig) {
+            leeWay = rotationConfig.indexLifetimeHard().minus(rotationConfig.indexLifetimeSoft());
+        } else {
+            throw new IllegalStateException(f("Unsupported RotationStrategyConfig type <%s>", indexSet.getConfig().rotationStrategy()));
+        }
+
+        if (indexExceedsSizeLimit(sizeInBytes)) {
+            return new Result(true,
+                    f("Index size <%s> exceeds MAX_INDEX_SIZE <%s>",
+                            StringUtils.humanReadableByteCount(sizeInBytes), MAX_INDEX_SIZE));
+        }
+        if (indexExceedsLeeWay(creationDate, leeWay)) {
+            return new Result(true,
+                    f("Index creation date <%s> exceeds optimization leeway <%s>",
+                            creationDate, leeWay));
+        }
+
+        if (indexIsOldEnough(creationDate) && !indexSubceedsSizeLimit(sizeInBytes)) {
+            return new Result(true,
+                    f("Index is old enough (<%s>) and has a reasonable size (<%s>) for rotation",
+                            creationDate, StringUtils.humanReadableByteCount(sizeInBytes)));
+        }
```

Comment:
We could optimize by only computing time-based limits when we are on a `ROTATION_PERIOD` boundary.
Not sure if the small amount of time-saving justifies the extra complexity though.

Reply:
> Do we want to snap to a grid of size ROTATION_PERIOD to ensure predictable times for rotation

yes, I didn't implement that yet. It's less trivial than I thought :-D

Let's postpone this and do it in the next PR.